cg/apps/downsample/downsample.py

```diff
@@ -0,0 +1,169 @@
+"""API that handles downsampling of samples."""
+import logging
+from pathlib import Path
+
+from cg.apps.downsample.utils import case_exists_in_statusdb, sample_exists_in_statusdb
+from cg.apps.housekeeper.hk import HousekeeperAPI
+from cg.constants import SequencingFileTag
+from cg.exc import DownsampleFailedError
+from cg.meta.meta import MetaAPI
+from cg.meta.workflow.downsample.downsample import DownsampleWorkflow
+from cg.models.cg_config import CGConfig
+from cg.models.downsample.downsample_data import DownsampleData
+from cg.store import Store
+from cg.store.models import Family, FamilySample, Sample
+from cg.utils.calculations import multiply_by_million
+
+LOG = logging.getLogger(__name__)
+
+
+class DownsampleAPI(MetaAPI):
+    def __init__(
+        self,
+        config: CGConfig,
+        sample_id: str,
+        number_of_reads: float,
+        case_id: str,
```

Comment:
I have a suggestion regarding the design pattern used:

Currently, the API is instantiated with specific details for each sample/case (e.g., sample_id, number_of_reads, case_id). This approach ties each instance of the API to a particular sample/case.

To enhance the API's usability and maintainability, I recommend decoupling the instance creation from specific samples/cases. Instead, we could:

1. Initialize DownsampleAPI with only configuration and shared resources, not tied to any sample or case.
2. Modify methods like downsample_sample to accept sample_id, number_of_reads, case_id, etc., as parameters. This way, one API instance can flexibly handle various samples and cases.
3. Ensure the API methods remain stateless regarding sample/case details, relying solely on the parameters provided to them.

This approach would make the API more general and versatile, simplifying the handling of different downsample tasks with a single, reusable API instance. Concretely, it would allow us to follow the pattern of only passing API:s via the context and also minimising the amount of logic in the CLI layer.

Reply:
Yeah, I think it is a bit more common - API:s usually provide a general set of methods.
Sure! I think the pattern I described is what generally applied throughout the cg already. We could document it in a design decision

Reply:
I like the approach you describe. I will give it a go!

Reply:
@seallard  have a look, I tried to implement a sample agnostic approach to the API. What do you think? Overall I can see the benefits of maintainability and reusability.